browser/brave_shields/ad_block_service_browsertest.cc

```diff
@@ -610,6 +613,88 @@ IN_PROC_BROWSER_TEST_F(AdBlockServiceTest,
   EXPECT_EQ(browser()->profile()->GetPrefs()->GetUint64(kAdsBlocked), 1ULL);
 }
 
+// Make sure that CNAME cloaked network requests get blocked correctly and
+// issue the correct number of DNS resolutions
+IN_PROC_BROWSER_TEST_F(AdBlockServiceTest, CnameCloakedRequestsGetBlocked) {
+  UpdateAdBlockInstanceWithRules("||cname-cloak-endpoint.tracking.com^");
+  EXPECT_EQ(browser()->profile()->GetPrefs()->GetUint64(kAdsBlocked), 0ULL);
+  GURL tab_url = embedded_test_server()->GetURL("a.com", kAdBlockTestPage);
+  GURL direct_resource_url =
+      embedded_test_server()->GetURL("a83idbka2e.a.com", "/logo.png");
+  GURL chain_resource_url =
+      embedded_test_server()->GetURL("b94jeclb3f.a.com", "/logo.png");
+  GURL safe_resource_url = embedded_test_server()->GetURL("a.com", "/logo.png");
+  GURL bad_resource_url = embedded_test_server()->GetURL(
+      "cname-cloak-endpoint.tracking.com", "/logo.png");
+
+  auto inner_resolver = std::make_unique<net::MockHostResolver>();
+
+  const std::vector<std::string> kDnsAliasesDirect(
+      {"cname-cloak-endpoint.tracking.com"});
+  const std::vector<std::string> kDnsAliasesChain(
+      {"cname-cloak-endpoint.tracking.com",
+       "cname-cloak-endpoint.tracking.com.redirectservice.net", "cname.a.com"});
+  const std::vector<std::string> kDnsAliasesSafe({"assets.cdn.net"});
+  inner_resolver->rules()->AddIPLiteralRuleWithDnsAliases(
+      "a83idbka2e.a.com", "127.0.0.1", kDnsAliasesDirect);
+  inner_resolver->rules()->AddIPLiteralRuleWithDnsAliases(
+      "b94jeclb3f.a.com", "127.0.0.1", kDnsAliasesChain);
+  inner_resolver->rules()->AddIPLiteralRuleWithDnsAliases("a.com", "127.0.0.1",
+                                                          kDnsAliasesSafe);
+  inner_resolver->rules()->AddIPLiteralRuleWithDnsAliases(
+      "cname-cloak-endpoint.tracking.com", "127.0.0.1", {});
+
+  network::HostResolver resolver(inner_resolver.get(), net::NetLog::Get());
+
+  brave::SetAdblockCnameHostResolverForTesting(&resolver);
```

Comment:
you also need to unset it somehow, so future tests would not get affected, see for example `ScopedBrowserLocale` or something like that

Reply:
good point! I've updated the test to reset it to `nullptr` again at the end, as well as some comments emphasizing that it's necessary. In the long-term some scoped solution would be ideal so I'll file a followup for that.

Reply:
Filed https://github.com/brave/brave-browser/issues/15619.